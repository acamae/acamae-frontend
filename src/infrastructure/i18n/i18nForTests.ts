import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';

// Import locales
import enGB from './locales/en-GB.json';
import esES from './locales/es-ES.json';

// Configure i18n for tests
i18n.use(initReactI18next).init({
  lng: 'en-GB',
  fallbackLng: 'en-GB',
  debug: false,
  resources: {
    'en-GB': {
      translation: enGB,
    },
    'es-ES': {
      translation: esES,
    },
  },
  interpolation: {
    escapeValue: false,
  },
});

export default i18n;
