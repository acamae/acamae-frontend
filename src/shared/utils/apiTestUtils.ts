import apiService from '@shared/services/axiosService';

/**
 * Represents the structured body of an API response, as defined by the application.
 */
export interface ApiResponseBody<TData = null> {
  success: boolean;
  data?: TData;
  message?: string;
  errors?: { field: string; errorKey: string }[];
}

// Simula la estructura de error de Axios para los mocks
export interface MockAxiosError extends Error {
  isAxiosError: boolean;
  response: {
    data: unknown;
    status: number;
  };
}

// Esta interfaz se exporta para que los tests puedan tipar sus par√°metros si lo desean
export interface SimulateApiCallParams<TResponseData = unknown, TRequestData = unknown> {
  method: 'get' | 'post' | 'put' | 'delete';
  url: string;
  requestData?: TRequestData;
  httpStatusCode: number;
  apiResponseBody: ApiResponseBody<TResponseData>;
  networkError?: Error; // Para simular errores donde no hay respuesta HTTP
}

export const setupMockApiServiceCall = <TResponseData = unknown, TRequestData = unknown>(
  params: SimulateApiCallParams<TResponseData, TRequestData>
) => {
  const {
    method,
    url: expectedUrl,
    requestData: expectedRequestData,
    httpStatusCode,
    apiResponseBody,
    networkError,
  } = params;

  const serviceMethod = (apiService as jest.Mocked<typeof apiService>)[method];

  if (networkError) {
    serviceMethod.mockImplementation(async (url: string, requestPayload?: unknown) => {
      if (url !== expectedUrl) {
        return Promise.reject(
          new Error(
            `Mocked ${method} (network error sim) called with unexpected URL: ${url}. Expected: ${expectedUrl}`
          )
        );
      }
      if (
        (method === 'post' || method === 'put') &&
        expectedRequestData &&
        JSON.stringify(requestPayload) !== JSON.stringify(expectedRequestData)
      ) {
        return Promise.reject(
          new Error(
            `Mocked ${method} (network error sim) to ${url} called with unexpected data. Expected: ${JSON.stringify(expectedRequestData)}, Got: ${JSON.stringify(requestPayload)}`
          )
        );
      }
      return Promise.reject(networkError);
    });
    return;
  }

  serviceMethod.mockImplementation(async (url: string, requestPayload?: unknown) => {
    if (url !== expectedUrl) {
      return Promise.reject(
        new Error(
          `Mocked ${method} (status ${httpStatusCode}) called with unexpected URL: ${url}. Expected: ${expectedUrl}`
        )
      );
    }
    if (
      (method === 'post' || method === 'put') &&
      expectedRequestData &&
      JSON.stringify(requestPayload) !== JSON.stringify(expectedRequestData)
    ) {
      return Promise.reject(
        new Error(
          `Mocked ${method} (status ${httpStatusCode}) to ${url} called with unexpected data. Expected: ${JSON.stringify(expectedRequestData)}, Got: ${JSON.stringify(requestPayload)}`
        )
      );
    }

    if (httpStatusCode >= 200 && httpStatusCode < 300) {
      return Promise.resolve({
        data: apiResponseBody,
        status: httpStatusCode,
      });
    } else if (httpStatusCode >= 300 && httpStatusCode < 600) {
      const error: MockAxiosError = Object.assign(
        new Error(`Request failed with status code ${httpStatusCode}`),
        {
          isAxiosError: true,
          response: {
            data: apiResponseBody,
            status: httpStatusCode,
          },
        }
      );
      return Promise.reject(error);
    } else {
      return Promise.reject(
        new Error(
          `Unsupported HTTP status code ${httpStatusCode} in mock setup for URL ${expectedUrl}.`
        )
      );
    }
  });
};

export const promiseMock = (shouldResolve = true) => {
  return jest.fn(() =>
    shouldResolve ? Promise.resolve() : Promise.reject(new Error('mock error'))
  );
};
