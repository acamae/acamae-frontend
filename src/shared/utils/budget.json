{
  "categories": {
    "performance": 90,
    "accessibility": 90,
    "best-practices": 90,
    "seo": 90
  },
  "metrics": {
    "largest-contentful-paint": [2500, 4000],
    "first-input-delay": [100, 300],
    "cumulative-layout-shift": [0.1, 0.25],
    "first-contentful-paint": [1800, 3000],
    "interactive": [3800, 7300],
    "total-blocking-time": [200, 600],
    "speed-index": [3400, 5800]
  },
  "timing": {
    "server-response-time": 600
  }
}
