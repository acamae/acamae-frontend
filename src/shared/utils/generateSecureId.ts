/**
 * Genera un ID criptográficamente seguro
 * Usa crypto.randomUUID() si está disponible, o crypto.getRandomValues() como alternativa
 *
 * @param dependencies Dependencias externas inyectables para testing
 */
export function generateSecureId(
  dependencies = {
    cryptoObj: typeof crypto !== 'undefined' ? crypto : undefined,
    dateNow: Date.now,
    processHrtime: typeof process !== 'undefined' ? process.hrtime : undefined,
  }
): string {
  const { cryptoObj, dateNow, processHrtime } = dependencies;

  try {
    // Intentar usar crypto.randomUUID() (disponible en Node.js 14.17+ y navegadores modernos)
    if (cryptoObj && typeof cryptoObj.randomUUID === 'function') {
      return cryptoObj.randomUUID().replace(/-/g, '').slice(0, 9);
    }

    // Alternativa con crypto.getRandomValues() (disponible en Node.js y navegadores)
    if (cryptoObj && typeof cryptoObj.getRandomValues === 'function') {
      const array = new Uint8Array(9);
      cryptoObj.getRandomValues(array);
      return Array.from(array, byte => byte.toString(36))
        .join('')
        .slice(0, 9);
    }

    // Última opción: usar timestamp con process.hrtime para entropía adicional
    // Esto no es criptográficamente seguro pero proporciona mejor entropía que Math.random()
    const timestamp = dateNow();
    const hrtime = processHrtime ? processHrtime.bigint().toString(36) : timestamp.toString(36);
    return `${timestamp}_${hrtime}`.slice(-9);
  } catch {
    // Si todos los métodos fallan, usar un fallback basado en timestamp
    const fallbackTimestamp = dateNow();
    return `${fallbackTimestamp}`.slice(-9);
  }
}
