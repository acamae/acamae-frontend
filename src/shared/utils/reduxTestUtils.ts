import { configureStore } from '@reduxjs/toolkit';

import authReducer, { initialAuthState } from '@application/state/slices/authSlice';
import sessionTimerReducer, {
  initialSessionTimerState,
} from '@application/state/slices/sessionTimerSlice';
import type { RootState } from '@domain/types/redux';

export function makeTestStore(preloadedState?: Partial<RootState>) {
  return configureStore({
    reducer: {
      auth: authReducer,
      sessionTimer: sessionTimerReducer,
    },
    preloadedState: {
      auth: initialAuthState,
      sessionTimer: initialSessionTimerState,
      ...preloadedState,
    } as RootState,
    middleware: getDefaultMiddleware => getDefaultMiddleware({ serializableCheck: false }),
  });
}
