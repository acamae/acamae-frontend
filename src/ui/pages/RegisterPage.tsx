import React from 'react';
import { Container, Row, Col, Card } from 'react-bootstrap';
import { useTranslation } from 'react-i18next';

import RegisterForm from '@/ui/components/Forms/RegisterForm';

const RegisterPage: React.FC = () => {
  const { t } = useTranslation();

  return (
    <Container className="py-5">
      <Row className="justify-content-center">
        <Col xs={12} md={8} lg={6} xl={5}>
          <Card>
            <Card.Header className="text-center">
              <h2 className="mb-4 text-center" data-testid="register-title">
                {t('register.title')}
              </h2>
            </Card.Header>
            <Card.Body>
              <RegisterForm redirectTo="/login?registrationSuccess=true" />
            </Card.Body>
          </Card>
        </Col>
      </Row>
    </Container>
  );
};

export default RegisterPage;
