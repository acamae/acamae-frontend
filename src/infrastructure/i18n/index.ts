/* eslint-disable import/no-named-as-default-member, import/no-named-as-default */
import i18n from 'i18next';
import LanguageDetector from 'i18next-browser-languagedetector';
import { initReactI18next } from 'react-i18next';

import enGB from '@infrastructure/i18n/locales/en-GB.json';
import esES from '@infrastructure/i18n/locales/es-ES.json';

const resources = {
  'es-ES': { translation: esES },
  'en-GB': { translation: enGB },
};

i18n
  .use(LanguageDetector) // Detects language from browser
  .use(initReactI18next) // Integrates i18next with React
  .init({
    resources,
    fallbackLng: 'es-ES', // Default language if translation is not found
    supportedLngs: ['es-ES', 'en-GB'],
    interpolation: {
      escapeValue: false, // React already escapes values by default
    },
    detection: {
      order: ['localStorage', 'navigator'], // Detection order: first localStorage, then navigator
      caches: ['localStorage'], // Stores language preference in localStorage
    },
  });

export default i18n;
