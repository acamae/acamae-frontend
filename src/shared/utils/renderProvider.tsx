import { configureStore, Store } from '@reduxjs/toolkit';
import { render } from '@testing-library/react';
import { I18nextProvider } from 'react-i18next';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

import rootReducer from '@application/state/rootReducer';
import { RootState } from '@domain/types/redux';
import i18n from '@infrastructure/i18n';

type TestProviderOptions = {
  preloadedState?: Partial<RootState>;
  store?: Store;
  route?: string;
  i18nInstance?: typeof i18n;
  withRouter?: boolean;
  withI18n?: boolean;
};

export function createTestProviderFactory(defaults: Partial<TestProviderOptions> = {}) {
  return function renderWithProviders(ui: React.ReactElement, options: TestProviderOptions = {}) {
    const {
      preloadedState = {},
      store = configureStore({ reducer: rootReducer, preloadedState }),
      route = '/',
      i18nInstance = i18n,
      withRouter = true,
      withI18n = true,
    } = { ...defaults, ...options };

    let tree = <Provider store={store}>{ui}</Provider>;
    if (withI18n) {
      tree = <I18nextProvider i18n={i18nInstance}>{tree}</I18nextProvider>;
    }
    if (withRouter) {
      tree = <MemoryRouter initialEntries={[route]}>{tree}</MemoryRouter>;
    }
    return render(tree);
  };
}
