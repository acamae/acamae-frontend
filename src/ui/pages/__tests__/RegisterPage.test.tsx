import { screen } from '@testing-library/react';

import i18n from '@infrastructure/i18n';
import { createTestProviderFactory } from '@shared/utils/renderProvider';
import RegisterPage from '@ui/pages/RegisterPage';

jest.mock('@ui/components/Forms/RegisterForm', () => {
  const MockRegisterForm = () => <div data-testid="mock-register-form">Mocked Register Form</div>;
  return MockRegisterForm;
});

function renderRegisterPage() {
  const renderWithProviders = createTestProviderFactory();
  return renderWithProviders(<RegisterPage />);
}

describe('RegisterPage', () => {
  it('should render the title', () => {
    renderRegisterPage();
    expect(screen.getByTestId('register-title')).toHaveTextContent(i18n.t('register.title'));
  });

  it('should render the form', () => {
    renderRegisterPage();
    expect(screen.getByTestId('mock-register-form')).toBeInTheDocument();
  });

  it('snapshot render', () => {
    const { asFragment } = renderRegisterPage();
    expect(asFragment()).toMatchSnapshot();
  });
});
