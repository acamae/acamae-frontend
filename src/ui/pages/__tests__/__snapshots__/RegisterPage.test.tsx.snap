// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterPage should render snapshot correctly 1`] = `
<DocumentFragment>
  <div
    class="register"
    data-testid="register-page"
  >
    <main
      class="register-content"
    >
      <h1
        class="text-center"
        data-testid="register-page-title"
      >
        Crear cuenta
      </h1>
      <div
        data-testid="mock-register-form"
      >
        Mocked Register Form
      </div>
    </main>
  </div>
</DocumentFragment>
`;
